netty/src/main/java/io/grpc/netty/AbstractNettyHandler.java

```diff
@@ -108,4 +121,109 @@ private void sendInitialConnectionWindow() throws Http2Exception {
       ctx.flush();
     }
   }
+
+  @VisibleForTesting
+  void setAutoTuneFlowControl(boolean isOn) {
+    autoTuneFlowControlOn = isOn;
+  }
+
+  /**
+   * Class for handling flow control pinging and flow control window updates as necessary.
+   */
+  final class FlowControlPinger {
+
+    private static final int MAX_WINDOW_SIZE = 8 * 1024 * 1024;
+    private int pingCount;
+    private int pingReturn;
+    private boolean pinging;
+    private int dataSizeSincePing;
+    private float lastBandwidth;
+    private long lastPingTime;
+
+    public int payload() {
+      return BDP_MEASUREMENT_PING;
+    }
+
+    public int maxWindow() {
+      return MAX_WINDOW_SIZE;
+    }
+
+    public void onDataRead(int dataLength, int paddingLength) {
+      if (!autoTuneFlowControlOn) {
+        return;
+      }
+      if (!isPinging()) {
+        setPinging(true);
+        sendPing(ctx());
+      }
+      incrementDataSincePing(dataLength + paddingLength);
+    }
+
+    public void updateWindow() throws Http2Exception {
+      if (!autoTuneFlowControlOn) {
+        return;
+      }
+      pingReturn++;
+      long elapsedTime = (System.nanoTime() - lastPingTime);
+      float bandwidth = 0;
+      if (elapsedTime > 0) {
+        bandwidth = (getDataSincePing() * TimeUnit.SECONDS.toNanos(1)) / elapsedTime;
+      }
+      Http2LocalFlowController fc = decoder().flowController();
+      // Calculate new window size by doubling the observed BDP, but cap at max window
+      int targetWindow = Math.min(getDataSincePing() * 2, MAX_WINDOW_SIZE);
+      setPinging(false);
+      int currentWindow = fc.initialWindowSize(connection().connectionStream());
+      if (targetWindow > currentWindow && bandwidth > lastBandwidth) {
```

Comment:
Did we need a multiplier for bandwidth? I do see that is a bit harder because the previous window isn't necessarily twice the current window (note that the target window doesn't come into play).